lib/stores/S3Store.ts

```diff
@@ -374,6 +356,48 @@ export default class S3Store extends DataStore {
     this.cache.delete(id)
   }
 
+  private calcOptimalPartSize(size: number): number {
+    let optimalPartSize: number
+
+    // When upload is smaller or equal to PreferredPartSize, we upload in just one part.
+    if (size <= this.preferredPartSize) {
+      optimalPartSize = size
+    }
+    // Does the upload fit in MaxMultipartParts parts or less with PreferredPartSize.
+    else if (size <= this.preferredPartSize * this.maxMultipartParts) {
+      optimalPartSize = this.preferredPartSize
+      // Prerequisite: Be aware, that the result of an integer division (x/y) is
+      // ALWAYS rounded DOWN, as there are no digits behind the comma.
+      // In order to find out, whether we have an exact result or a rounded down
+      // one, we can check, whether the remainder of that division is 0 (x%y == 0).
+      //
+      // So if the result of (size/MaxMultipartParts) is not a rounded down value,
+      // then we can use it as our optimalPartSize. But if this division produces a
+      // remainder, we have to round up the result by adding +1. Otherwise our
+      // upload would not fit into MaxMultipartParts number of parts with that
+      // size. We would need an additional part in order to upload everything.
+      // While in almost all cases, we could skip the check for the remainder and
+      // just add +1 to every result, but there is one case, where doing that would
+      // doom our upload. When (MaxObjectSize == MaxPartSize * MaxMultipartParts),
+      // by adding +1, we would end up with an optimalPartSize > MaxPartSize.
+      // With the current S3 API specifications, we will not run into this problem,
+      // but these specs are subject to change, and there are other stores as well,
+      // which are implementing the S3 API (e.g. RIAK, Ceph RadosGW), but might
+      // have different settings.
+    } else if (size % this.maxMultipartParts === 0) {
+      optimalPartSize = size / this.maxMultipartParts
+      // Having a remainder larger than 0 means, the float result would have
+      // digits after the comma (e.g. be something like 10.9). As a result, we can
+      // only squeeze our upload into MaxMultipartParts parts, if we rounded UP
+      // this division's result. That is what is happending here. We round up by
+      // adding +1, if the prior test for (remainder == 0) did not succeed.
+    } else {
+      optimalPartSize = size / this.maxMultipartParts + 1
```

Comment:
I am concerned about this division as it might not return an integer but a decimal value. Maybe just use

```suggestion
      optimalPartSize = Math.floor(size / this.maxMultipartParts) + 1
```

and update the comments above.